src/components/geolocate-control.js

```diff
@@ -0,0 +1,285 @@
+      this[prop] = data[prop];
+    }
+
+    this.type = type;
+  }
+}
+
+const propTypes = Object.assign({}, BaseControl.propTypes, {
+  // Custom className
+  className: PropTypes.string,
+
+  // mapbox geolocate options
+  // https://docs.mapbox.com/mapbox-gl-js/api/#geolocatecontrol
+  positionOptions: PropTypes.object,
+  fitBoundsOptions: PropTypes.object,
+  trackUserLocation: PropTypes.bool,
+  showUserLocation: PropTypes.bool,
+
+  // Callbacks fired when the user interacted with the map. The object passed to the callbacks
+  // contains viewport properties such as `longitude`, `latitude`, `zoom` etc.
+  onViewStateChange: PropTypes.func,
+  onViewportChange: PropTypes.func
+});
+
+const defaultProps = Object.assign({}, BaseControl.defaultProps, {
+  className: '',
+  // mapbox geolocate options
+  positionOptions: {
+    enableHighAccuracy: false,
+    maximumAge: 0,
+    timeout: 6000 /* 6 sec */
+  },
+  fitBoundsOptions: {
+    maxZoom: 15
+  },
+  trackUserLocation: false,
+  showUserLocation: true,
+
+  // viewport handlers
+  onViewStateChange: () => {},
+  onViewportChange: () => {}
+});
+
+const WATCH_STATE = {
+  OFF: 'OFF',
+  WAITING: 'WAITING',
+  WAITING_ACTIVE: 'WAITING_ACTIVE',
+  ACTIVE: 'ACTIVE',
+  ACTIVE_LOCK: 'ACTIVE_LOCK',
+  ACTIVE_ERROR: 'ACTIVE_ERROR',
+  BACKGROUND: 'BACKGROUND',
+  BACKGROUND_ERROR: 'BACKGROUND_ERROR'
+};
+
+export default class GeolocateControl extends BaseControl {
+  static propTypes = propTypes;
+  static defaultProps = defaultProps;
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      supportsGeolocation: undefined,
+      markerPosition: {
+        longitude: 0,
+        latitude: 0
+      }
+    };
+  }
+
+  _containerRef : { current: null | HTMLDivElement } = createRef();
+  _geolocateButtonRef : { current: null | HTMLDivElement } = createRef();
+  _markerRef : { current: null | HTMLDivElement } = createRef();
+
+  componentDidMount() {
+    this._checkGeolocationSupport();
+  }
+
+  _setupMapboxGeolocateControl = (supportsGeolocation) => {
+    if (!supportsGeolocation) {
+      /* eslint-disable no-console, no-undef */
+      console.warn(
+        'Geolocation support is not available, the GeolocateControl will not be visible.'
+      );
+      /* eslint-enable no-console, no-undef */
+      return;
+    }
+
+    const controlOptions = {};
+    ['positionOptions', 'fitBoundsOptions', 'trackUserLocation', 'showUserLocation'].forEach(
+      prop => {
+        if (this.props[prop]) {
+          controlOptions[prop] = this.props[prop];
+        }
+      }
+    );
+
+    this._mapboxGeolocateControl = new MapboxGeolocateControl(controlOptions);
+
+    // replace mapbox internal prop
+    this._mapboxGeolocateControl._map = this._context.map;
+    this._mapboxGeolocateControl._watchState = WATCH_STATE.OFF;
+
+    // replace mapbox internal UI elements
+    this._mapboxGeolocateControl._geolocateButton = this._geolocateButtonRef.current;
```

Comment:
We should render the UI components in React.

Reply:
Inside mapbox's `geolocate-control`, all the functions related to handle locating user, `trigger, _onSuccess, _onError`, changed the `classList` of `_geolocateButton`, so I need replace the  mapbox's `_geolocateButton` with our own react button element, then mapbox can manipulate the css of `_geolocateButton`. 

I did this because I think we should use Mapbox's `trigger, _onSuccess, _onError` instead of duplicate the code.